Limb grower

hamzah-hayat opened the pull request:
:cl:
add: Added a Experimental Limb Grower to the medbay department, now medical stuff can grow synthetic limbs using Synthflesh to help crew that suffer any work related accidents. There may also be a limb that can be grown with a special card...
/:cl:

As promised, I've fully finished and fleshed out this limb grower, and i've added it to both meta station and box station.
Credits for the sprite goes to my sister.

hamzah-hayat committed:
Added an AltClick function to split Stacks

Stacks now have an altclick function that lets you input how much you
would like to split the stack by

Also changed the regular click code by turning most of its code into a
proc, this reduced the amount of copy-pasting needed, as both the
regular and alt click on stacks do the same thing, except alt clicking
lets you input a number

hamzah-hayat committed:
Wrong var name

hamzah-hayat committed:
Edits based on feedback

-Removed trailing return
-removed uneeded if statement
-removed uneeded src as it should call the right method anyway

hamzah-hayat committed:
Removed metal interact after splitting

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into tgstation/master

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into tgstation/master

hamzah-hayat committed:
Adds basic limb grower to game

hamzah-hayat committed:
Added includes

hamzah-hayat committed:
Can now add synthflesh to grower

Limbgrower can take in synthflesh, starts with 100
uses glass beakers to construct, also changed tech for testing
removed a lot of unneeded code left over from autolathe, including
references to materials

hamzah-hayat committed:
Bugfixes and added to boxstation

fixes bug where limb grower starts with empty reagent container (with
nothing inside it)
also trying to fix limb creating bug

Also added limbgrower to boxstation for easier testing

hamzah-hayat committed:
changed where limb spawns and costs

hamzah-hayat committed:
Fixed menus, added sprites

Also cleaned up all the unneeded code

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

hamzah-hayat committed:
Merge master into Limbgrower

hamzah-hayat committed:
Merge

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Limb-Grower

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Limb-Grower

hamzah-hayat committed:
Fixed the map

hamzah-hayat committed:
Probally the cause of the build failure

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Limb-Grower

hamzah-hayat committed:
trying to fix the build

hamzah-hayat committed:
Limbs are now produced properly

Limbs made properly depending on specias
added plasman as a limb option
addedd a check for beakers putting synthflesh into the grower

hamzah-hayat committed:
Fixed menus, added emag, changed balance

hamzah-hayat committed:
Added the Limbgrower to meta and box station

hamzah-hayat committed:
Got rid of debug messages, fixed emag

also limb grower starts with less synthflesh

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Limb-Grower

hamzah-hayat committed:
changed menu, builds armbaldes (i think)

minor map edit

hamzah-hayat committed:
Fixed spelling, added synthetic arm blade

this way arm blades dont attach themselves on equip, and also they arent
quite as amazing as the real thing

hamzah-hayat committed:
fixes travis build

phil235 commented:
The code never uses that 'human' var.

Jalleo commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,264 @@
+			if(reagents.has_reagent("synthflesh", being_built.reagents["synthflesh"]*prod_coeff))
+				busy = 1
+				use_power(power)
+				flick("limbgrower_fill",src)
```
Flick is bad.

optimumtact replied:
why exactly do you think flick is bad for this?

chickenboy10 replied:
I use the flick procs here to run a single use animation of the limb grower "filling" up, and another when i want it to "fill" down.
![fillup](https://cloud.githubusercontent.com/assets/11432022/18494218/dad5340c-7a0c-11e6-9e63-c44ee67ed602.gif)
![filldown](https://cloud.githubusercontent.com/assets/11432022/18494219/dd924a36-7a0c-11e6-98e4-e3a1d956531c.gif)

while its in the middle of making the limb i keep it on this as its iconstate 
![idle](https://cloud.githubusercontent.com/assets/11432022/18494238/f11e0842-7a0c-11e6-8c99-71cda86e2007.gif)

so i think its fine to use flick here

Jalleo replied:
Its because last I knew flick is a server side animation where as the animate proc is clienside. Which if I remember correctly means all clients get the flick activate. Whereas animate is just the ones that see it.

Jalleo commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,264 @@
+					use_power(power)
+					reagents.remove_reagent("synthflesh",being_built.reagents["synthflesh"]*prod_coeff)
+					var/B = being_built.build_path
+					if(ispath(B, /obj/item/bodypart))	//This feels like spatgheti code, but i need to initilise a limb somehow
```
I think there is something that @Remierichards mentioned about this.  But i cant remember off hand and on a mobile.

optimumtact commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,264 @@
+/obj/machinery/limbgrower/attackby(obj/item/O, mob/user, params)
+	if (busy)
+		user << "<span class=\"alert\">The Limb Grower is busy. Please wait for completion of previous operation.</span>"
+		return 1
```
why do some returns return 1 and then others return null?

Should it not be 1 and 0, and if so, use the TRUE/FALSE defines

optimumtact commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,264 @@
+				use_power(power)
+				flick("limbgrower_fill",src)
+				icon_state = "limbgrower_idleon"
+				spawn(32*prod_coeff)
```
avoid this spawn, either use the timer subsystem https://github.com/tgstation/tgstation/blob/master/code/controllers/subsystem/timer.dm#L65

or refactor this to use world.time + cooldown and then handling the actual item part building in the process proc of this machine (don't forget to call the parent process with ..())

optimumtact commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,264 @@
+					busy = 0
+					flick("limbgrower_unfill",src)
+					icon_state = "limbgrower_idleoff"
+					src.updateUsrDialog()
```
src should not be needed here, same with a few other locations.

hamzah-hayat committed:
bugfixes

- removed starting synthflesh, small beaker spawns ontop of the machine
with synthflesh on the maps
- got rid of human variable
-cleaned up return values and other bits and pieces
-added synthflesh beaker

optimumtact commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,255 @@
+		build_limb(B)
+	else
+		//Just build whatever it is
+		var/obj/L = new B()
```
instead of doing L.loc = loc; below, you can do

    var/obj/L = new B(loc)

optimumtact commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,255 @@
+	//i need to create a body part manually using a set specias dna
+	var/obj/item/bodypart/L
+	var/selected_part
+	L = new B()
```
same here you can just do

     L = new B(loc)

optimumtact commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,255 @@
+	return dat
+
+/obj/machinery/limbgrower/emag_act(mob/user)
+	if(emag==1)
```
use the define

     emag == TRUE

and

    emag = TRUE

It's a little bit clearer that the value is a boolean (sort of) flag

hamzah-hayat committed:
more refactoring

sadly using the timer over spawn has broken the machine, it now waits
infinitely and never finishes, need to fix

optimumtact commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,253 @@
+	return
+
+/obj/machinery/limbgrower/proc/build_item()
+	reagents.remove_reagent("synthflesh",being_built.reagents["synthflesh"]*prod_coeff)
```
you may want to check that it has enough reagent, as this call happens after time has passed (and reagent amount may have changed in the mean time)

chickenboy10 replied:
I dont think that possibly could happen, but ill add a check just in case

optimumtact replied:
it's usually a case of outside parties fiddling, i.e admins removing reagents, but it can happen

hamzah-hayat committed:
bugfixes

- turns out i cant spell build_item
- sanity check just in case the reagents change between start and finish
of building

hamzah-hayat committed:
spelt usr wrong

its almost 4am

optimumtact commented:
Look good to me :+1:

chickenboy10 commented:
cool, thanks, its getting late and I really wanted to finish this before going to sleep.
I think the travis build will complain because I don't ever use L.
code\game\machinery\limbgrower.dm:147:warning: L: variable defined but not used

hamzah-hayat committed:
Dont need L

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Limb-Grower

phil235 commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,256 @@
+		return
+
+	if(O.flags & HOLOGRAM)
+		return
```
the last four lines do nothing. And the harm check is only useful for machines that requires things to be inserted in them.

phil235 commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,256 @@
+		selected_part = "left leg"
+		L.icon_state = "[selected_category]_l_leg_s"
+	L.name = "Synthetic [selected_category] Limb"
+	L.desc = "A synthetic [selected_category] limb that will morph on its first use in surgery. This one is for the [selected_part]"
```
These limbs are useless, can't use them for surgery, because they don't have the required vars to match a right leg.

chickenboy10 replied:
I've tested them out and they attach fine to mobs. They are just regular arms and legs with the "dropped" value set to false. I set them a icon because otherwise they would be invisible (but still work) as they havent been "dropped" from a mob so dont know what they should look like.

phil235 replied:
See https://github.com/tgstation/tgstation/blob/master/code/modules/surgery/prosthetic_replacement.dm#L39

phil235 replied:
Also, you can remove selected_part entirely and use [parse_zone(L.body_zone)] directly in the desc.

hamzah-hayat committed:
refactor and rebalance

limbs are cheaper now

hamzah-hayat committed:
what

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/origin/Limb-Grower'

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/origin/StackSplit'

hamzah-hayat committed:
No longer uses "z" as part of the code

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

hamzah-hayat committed:
Merge branch 'refs/heads/abandoned-crate-fix'

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Limb-Grower

hamzah-hayat committed:
shoudnt be here

pudl commented:
pls get a new sprite

phil235 commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,248 @@
+	icon_state = "limbgrower_idleoff"
+	updateUsrDialog()
+
+/obj/machinery/limbgrower/proc/build_limb(var/B)
```
the "var/" in the argument is not needed. Also "B" is way too generic for an argument name, it's usually best to give arguments names that describe their function, anything that helps the reader understand what this argument is about.

phil235 commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,248 @@
+/obj/machinery/limbgrower/proc/build_item()
+	if(reagents.has_reagent("synthflesh", being_built.reagents["synthflesh"]*prod_coeff))	//sanity check, if this happens we are in big trouble
+		reagents.remove_reagent("synthflesh",being_built.reagents["synthflesh"]*prod_coeff)
+		var/B = being_built.build_path
```
find a better var name, like var/limb_path

phil235 commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,248 @@
+			build_limb(B)
+		else
+			//Just build whatever it is
+			new B(loc)
```
is there any designs used by this machine that actually is not a path for a limb?

phil235 commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,248 @@
+		L.icon_state = "[selected_category]_r_leg_s"
+	else if(ispath(B,/obj/item/bodypart/l_leg))
+		selected_part = "left leg"
+		L.icon_state = "[selected_category]_l_leg_s"
```
You can simplify the code by using `L.icon_state = "[selected_category]_[L.body_zone]_s"`. This way you don't need all those ispath() checks.

phil235 commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,248 @@
+	else if(ispath(B,/obj/item/bodypart/l_leg))
+		selected_part = "left leg"
+		L.icon_state = "[selected_category]_l_leg_s"
+	L.name = "Synthetic [selected_category] Limb"
```
`L.name = "Synthetic [selected_category] [parse_zone(L.body_zone)]"`
And since you'll use parse_zone(L.body_zone) again for the desc below you can make it into a var so we don't have to call the parse_zone() proc twice.

hamzah-hayat committed:
More refactoring

Thanks for help, makes it much easier to build limbs now.

hamzah-hayat committed:
Forgot these two lines

GunHog commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,238 @@
+	return dat
+
+/obj/machinery/limbgrower/emag_act(mob/user)
+	if(emag==TRUE)
```
`if(emag)`

phil235 commented on code/game/machinery/limbgrower.dm:
```diff
@@ -131,41 +131,29 @@
+	else
+		limb.icon = 'icons/mob/human_parts.dmi'
+	// Set this limb up using the specias name and body zone
+	limb.icon_state = "[selected_category]_[limb.body_zone]_s"
```
you removed the name and desc change.

chickenboy10 replied:
yep i accidently forgot it, added it back in just now

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master' into Limb-Grower

optimumtact commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,238 @@
+	for(var/datum/design/D in files.possible_designs)
+		if((D.build_type & LIMBGROWER) && ("special" in D.category))
+			files.AddDesign2Known(D)
+	usr << "A warning flashes onto the screen, stating that safety overrides have been deactivated"
```
use user here, not usr, we try to avoid using usr unless absolutely necessary

optimumtact commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,238 @@
+			//Just build whatever it is
+			new buildpath(loc)
+	else
+		usr << "<span class=\"error\"> Something went very wrong and there isnt enough synthflesh anymore!</span>"
```
this should use src.visible_message

optimumtact replied:
because a usr may no longer exist for the machine

hamzah-hayat committed:
refactor and hopefully merge fix

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master' into new-Master

hamzah-hayat committed:
Merge branch 'refs/heads/new-Master' into Limb-Grower

hamzah-hayat committed:
readded machine define

for limbgrower

Chocobro commented:
that sprite is super cool, give your sister my props :+1:
